### ISSUE TITLE
[Bug]: DGX Spark GB10 repeatedly loses host/network reachability during single-node multimodal heldout evaluation

### ISSUE THREAD

**Issue by AuricTW** [NVIDIA/dgx-spark-playbooks#116]
> **Correction (2026-08-31):** The initial version of this report incorrectly described the two recent host-loss incidents as QLoRA training runs and implied that both had temporally adjacent RM allocation errors. Artifact timestamps and boot journals show that both incidents occurred during single-node vLLM heldout evaluation. The QLoRA run was a separate, later controlled memory-pressure experiment. The report below has been corrected to keep those evidence sets separate.

## Summary

A single NVIDIA DGX Spark / GB10 lost network, SSH and all monitoring targets twice during two separate single-node vLLM multimodal heldout evaluations. The runs used different fine-tuned versions of the same approximately 2B vision-language model:

1. an answer-only, QLoRA-derived model
2. a full-rank MixGOLD model

Both evaluations completed the first eight 128-sample tasks and became unreachable partway through the ninth task, DocVQA. Neither run produced a final summary or exit code, and each DocVQA JSONL ended with one invalid, NUL-filled trailing record.

The two incidents are repeatable at the workload-stage level, but the available evidence does **not** establish a common low-level root cause:

- the first incident's boot had RM `NV_ERR_NO_MEMORY` messages approximately 12.5 hours earlier, not near the outage
- the second incident's boot had no NVRM allocation error at all
- the final Prometheus samples still showed approximately 45.5 GiB and 43.4 GiB `MemAvailable`
- no Linux OOM-killer event, NVIDIA Xid, thermal alarm, recorded panic or orderly shutdown was found near either event

A separate controlled PyTorch QLoRA experiment later reproduced `_memdescAllocInternal` and IOVA metadata allocation failures under severe UMA pressure and found a deterministic `torch.cuda.empty_cache()` mitigation for that training workload. That experiment is included as related evidence, not as proof that it caused the two heldout outages.

## Expected behavior

A benchmark process or GPU runtime failure should be contained to the process and return an actionable error. It should not make the entire DGX Spark disappear from the network without a recorded panic, Xid or orderly shutdown.

## Environment

- Hardware: NVIDIA DGX Spark, NVIDIA GB10, single node
- OS: Ubuntu 24.04.4 LTS, aarch64
- Kernel: `6.17.0-1014-nvidia #14-Ubuntu`
- NVIDIA driver: `580.142`
- CUDA: `13.0`
- Evaluation runtime: single-node vLLM
- Python: `3.10.20`
- PyTorch: `2.10.0`
- Transformers: `5.6.2`

## Heldout workload

- nine benchmark families
- 128 samples per task, 1,152 expected rows
- deterministic task and sample order
- `temperature=0`
- `max_tokens=4096`
- maximum 8 video frames
- DocVQA was the ninth and final task

The two affected models used different adaptation methods, which makes a QLoRA-training-only explanation unlikely:

| Incident | Evaluated model | Completed valid rows | DocVQA valid rows | Final marker |
|---|---|---:|---:|---|
| A | Answer-only 4X, QLoRA-derived | 1,119 / 1,152 | 95 / 128 | no summary or exit code; one invalid trailing JSONL record |
| B | MixGOLD 4X, full-rank BF16 | 1,113 / 1,152 | 89 / 128 | no summary or exit code; one invalid trailing JSONL record |

In both cases, the preceding eight tasks had exactly 128/128 valid rows. The last valid DocVQA sample was different, so the evidence identifies a repeatable stage rather than one deterministic sample.

## Host-loss timing and telemetry

Prometheus CPU data below uses an `up == 1` mask so target-down periods cannot create a misleading graph tail.

| Metric | Incident A | Incident B |
|---|---:|---:|
| Heldout start | 00:32:05 | 18:49:15 |
| Last node target up | approximately 00:52:15 | 19:09:35 |
| First node target down | 00:52:30 | 19:09:40 |
| Minimum sampled `MemAvailable` in the evaluation window | 44.28 GiB | 43.41 GiB |
| Last sampled `MemAvailable` | 45.51 GiB | 43.41 GiB |
| Corrected CPU maximum | 49.07% | 46.46% |
| Corrected CPU last sample | 9.51% | 10.16% |
| Maximum load1 | 14.84 | 13.18 |
| Maximum ACPI thermal reading | 92.7 C | 93.7 C |
| Last ACPI thermal reading | 91.2 C | 91.9 C |

These samples do not support ordinary host-RAM exhaustion, sustained CPU saturation or a simple thermal-threshold shutdown. They cannot exclude a transient event between scrapes or a driver/firmware/platform failure that stopped logging.

### Incident A

- The boot journal stopped at 00:50:58; the node target stopped responding shortly afterward.
- No Linux OOM-killer, Xid, thermal alarm, panic or orderly shutdown was recorded near the event.
- The same boot did contain `_memdescAllocInternal` `NV_ERR_NO_MEMORY` messages at approximately 12:15 on the previous day, during a different workload. The gap is over 12 hours, so they cannot be treated as a temporally adjacent cause of this heldout outage.

### Incident B

- The boot journal stopped at 19:08:25; the node target changed to down at 19:09:40.
- No NVRM allocation error occurred anywhere in this boot.
- No Linux OOM-killer, Xid, thermal alarm, panic or orderly shutdown was recorded near the event.

The last journal records include local SSH health-check and network-interface activity, but they do not establish whether the primary failure was the network stack, GPU/driver, firmware, watchdog or another platform component.

## Separate controlled RM/IOVA experiment

After the heldout incidents, we used a checkpointed PyTorch/Transformers multimodal QLoRA workload to test a separate UMA-memory-pressure hypothesis. This was not the workload that lost the host in incidents A or B.

The controlled workload used:

- an NF4 4-bit approximately 2B student
- a frozen BF16 reference model
- LoRA rank 16 on `q_proj`, `k_proj`, `v_proj`, `o_proj`
- answer CE + rationale CE + forward KL
- 4,608 heterogeneous image/video samples
- a safety gate at `MemAvailable < 16 GiB`

During a checkpoint resume, `MemAvailable` fell from approximately 34.0 GiB to 12.7 GiB in about 1.5 seconds. The safety gate sent SIGTERM at 14:02:49.395749. The first RM line's own kernel timestamp was 14:02:49.402081, approximately 6.3 ms later:

```text
NVRM: nvCheckOkFailedNoLog: Check failed: Out of memory [NV_ERR_NO_MEMORY]
(0x00000051) returned from _memdescAllocInternal(pMemDesc) @ mem_desc.c:1359

NVRM: _iovaspaceCreateMappingDataFromMemDesc: failed to allocate
0x176028 bytes for IOVA mapping metadata
```

Because SIGTERM came first, this controlled result does not prove whether the failed RM operation was an allocation already in flight or teardown under severe memory pressure.

An A/B then changed only one behavior: call `torch.cuda.empty_cache()` between completed training samples.

| Controlled training result | No per-sample cache release | Per-sample cache release |
|---|---:|---:|
| Remaining samples completed | 117 / 512 before safety stop | 512 / 512 |
| Maximum recorded CUDA reserved in a complete reference / A/B | 112.777 GiB | 89.812 GiB |
| New RM errors in the controlled telemetry window | 3 | 0 |
| Per-sample loss difference versus complete reference | n/a | exactly 0.0 |
| Final adapter | no final adapter from stopped run | byte-identical to complete reference |

This is strong evidence that retained CUDA cache materially contributed to the controlled QLoRA memory-pressure boundary, and that cache release is a practical mitigation for that workload. It is not evidence that the two vLLM heldout host losses had the same cause.

## Additional controls

- Five fresh-engine DocVQA repetitions completed 128/128 each without an RM error.
- A separate continuous nine-task heldout completed 9 x 128 samples without an RM error; minimum `MemAvailable` was 41.0 GiB.
- Historical journals contain many `_memdescAllocInternal` failures from other high-memory workloads, including events that did not immediately cause host loss.

Therefore neither DocVQA alone nor an RM error alone is sufficient to reproduce the observed whole-host reachability loss.

## Current assessment

Confirmed:

- both recent host losses occurred during vLLM heldout evaluation, not training
- both became unreachable during the final DocVQA stage after the first eight tasks completed
- ordinary host-memory exhaustion, sustained CPU saturation, a Linux OOM kill and a simple temperature threshold are not supported by the sampled evidence
- the controlled QLoRA workload has a real, independently reproducible RM/IOVA memory-pressure problem that is mitigated by per-sample cache release

Not confirmed:

- that the two heldout outages share the same low-level cause
- that either heldout outage was caused by `_memdescAllocInternal`
- whether the failing component was vLLM, NVIDIA RM/UVM, the network stack, firmware, watchdog or another platform component

## Questions for NVIDIA

1. What diagnostics can distinguish a GPU/driver hard lock from a network-stack or platform failure when the journal stops without Xid, panic or shutdown records?
2. Can a GB10 RM/UVM or firmware failure make all network services unreachable while leaving no NVRM line in the affected boot?
3. Which additional tracing should be enabled for a repeatable single-node vLLM multimodal heldout run: UVM/RM debug logs, netconsole, pstore, watchdog tracing, kdump or another mechanism?
4. Is approximately 43-45 GiB OS `MemAvailable` sufficient to rule out a GB10 GPU-visible UMA/IOVA resource exhaustion state?
5. Is driver `580.142` known to have whole-host stability issues under mixed image/video vLLM workloads, and which current DGX Spark driver/kernel/firmware stack should be used for confirmation?

Kdump is now configured, but we have not deliberately forced another host loss. We can provide sanitized per-sample artifacts, Prometheus exports, boot journals, the controlled training wrapper and an `nvidia-bug-report.sh` bundle through an appropriate private channel.

## Related reports

- #97 reports the same RM call site from vLLM long-context and startup paths. Our separate controlled QLoRA experiment reached that call site, but the two heldout host-loss incidents do not have temporally adjacent RM evidence.
- vllm-project/vllm#46307 discusses GB10 unified-memory pressure during vLLM startup profiling.
